payment/app.py

```diff
@@ -12,39 +14,164 @@
+    except Exception as err:
+        return Response(str(err), status=400)
+
+    body = {
+        "done" : True
+    }
+
+    return Response(json.dumps(body), mimetype="application/json", status=200)
 
 @app.post('/pay/<user_id>/<order_id>/<amount>')
 def remove_credit(user_id: str, order_id: str, amount: int):
-    pass
+    
+    if int(amount) <= 0:
+        return Response("The amount must be >0!", status=400)
+
+    if not db.hget(user_id, "user_id"):
+        return Response(f"The user {user_id} does not exist in the DB!", status=404)
+
+    order_find_path = gateway_url + f"order-service/orders/find/{order_id}"
+    r = requests.get(order_find_path)
+    response_json = json.load(r.json())
+
+    if r.status_code==404:
+        return Response(f"The order {order_id} does not exist in the DB!", status=404)
 
+    if bool(response_json["paid"]):
+        return Response(f"The order {order_id} has been already paid", status=400)
+
+    current_credit = int(db.hget(user_id, "credit"))
+
+    if current_credit < int(amount):
+        return Response(f"Insufficient credit balance")
+
+    try:
+        new_credit = db.hincrby(user_id, "credit", -1*int(amount))
+    except Exception as err:
+        return Response(str(err), status=400)
+
+    return Response(f"The payment of the order {order_id} is paid", status=200)
 
 @app.post('/cancel/<user_id>/<order_id>')
 def cancel_payment(user_id: str, order_id: str):
-    pass
-
+    if not db.hget(user_id, "user_id"):
+        return Response(f"The user {user_id} does not exist in the DB!", status=404)
+    
+    order_find_path = gateway_url + f"order-service/orders/find/{order_id}"
+    r = requests.get(order_find_path)
+    response_json = json.load(r.json())
+
+    if r.status_code==404:
+        return Response(f"The order {order_id} does not exist in the DB!", status=404)
+
+    if not bool(response_json["paid"]):
+        return Response(f"The payment of the order {order_id} has not been made yet", status=400)
+
+    # Add item back to stock
+    item_list = response_json["items"]
+
+    for item in item_list:
+        try:
+            stock_add_path = gateway_url + f"stock-service/add{item}/1"
+            requests.post(stock_add_path)
+        except Exception as err:
+            return Response(str(err), status=400)
+
+    # Refund
+    refund_amount = int(response_json["total_cost"])
+    try:
+        db.hincrby(user_id, "credit", refund_amount)
+    except Exception as err:
+        return Response(str(err), status=400)
+
+    return Response(f"The order {order_id} is canceled successfully!", status=200)
 
 @app.post('/status/<user_id>/<order_id>')
 def payment_status(user_id: str, order_id: str):
-    pass
+    
+    if not db.hget(user_id, "user_id"):
+        return Response(f"The user {user_id} does not exist in the DB!", status=404)
+    
+    order_find_path = gateway_url + f"order-service/orders/find/{order_id}"
```

Comment:
Is this URL working correctly?

Reply:
These urls should be fixed to the correct ones but it can only be tested after merging.